compiler/optimizations/gpuTransforms.cpp

```diff
@@ -1140,85 +1205,92 @@ static void generateGpuAndNonGpuPaths(const GpuizableLoop &gpuLoop,
 
-    // then relocate the loop
-    cpuBlock->insertAtHead(gpuLoop.loop()->remove());
+  gpuLoop.gpuLoop()->replace(gpuBlock);
+}
 
-    // if not doing GPU codegen, just add cpuBlock after the conditional
-    if (!isFullGpuCodegen()) {
-      cond->insertAfter(cpuBlock);
+static CallExpr* getGpuEligibleMarker(CForLoop* loop) {
+  if (loop->body.length > 0) {
+    if (auto callExpr = toCallExpr(loop->body.get(1))) {
+      if (callExpr->primitive && callExpr->primitive->tag == PRIM_GPU_ELIGIBLE) {
+        return callExpr;
+      }
     }
   }
+  return nullptr;
 }
 
 static void outlineEligibleLoop(FnSymbol *fn, GpuizableLoop &gpuLoop) {
-  SET_LINENO(gpuLoop.loop());
+  SET_LINENO(gpuLoop.gpuLoop());
+
+  // The marker is a compile-time only primitive; remove it now.
+  if (auto marker = getGpuEligibleMarker(gpuLoop.gpuLoop())) {
+    marker->remove();
+  }
 
   // Construction of the GpuKernel will create the outlined function
   GpuKernel kernel(gpuLoop, fn->defPoint);
   if(!kernel.lateGpuizationFailure()) {
-    generateGpuAndNonGpuPaths(gpuLoop, kernel);
+    generateGPUKernelCall(gpuLoop, kernel);
   } else {
     kernel.fn()->defPoint->remove();
   }
 }
 
 static void outlineGpuKernelsInFn(FnSymbol *fn) {
-  std::vector<BaseAST*> asts;
-  collect_asts(fn, asts);
-
-  for_vector(BaseAST, ast, asts) {
-    if (CForLoop* loop = toCForLoop(ast)) {
-      // In the case of a nested foreach loop we may end up replacing the
-      // outer loop with a kernel call and in doing so making the loop no
-      // longer in the tree.
-      if (!loop->inTree()) {
-        continue;
-      }
+  std::vector<CForLoop*> asts;
+  collectCForLoopStmtsPreorder(fn, asts);
+
+  for_vector(CForLoop, loop, asts) {
+    // In the case of a nested foreach loop we may end up replacing the
+    // outer loop with a kernel call and in doing so making the loop no
+    // longer in the tree.
+    if (!loop->inTree()) {
+      continue;
+    }
 
-      GpuizableLoop gpuLoop(loop);
+    auto foundLoop = eligibleLoops.find(loop);
+    if (foundLoop != eligibleLoops.end()) {
+      // The loops originally in eligibleLoops should be in non-specialized
+      // procs, and thus, should not ever be encountered during outlining.
+      INT_ASSERT(!fGpuSpecialization);
+
+      // The GPU branches we create for eligible loops are always if-else,
+      // not if-fallthrough. Make them fall through if necessary (that is,
+      // if we're in CPU-as-device mode).
+      auto& eligibleLoop = foundLoop->second;
+      eligibleLoop.fixupNonGpuPath();
+      outlineEligibleLoop(fn, eligibleLoop);
+    } else if (fGpuSpecialization && getGpuEligibleMarker(loop)) {
+      // Even if this wasn't a loop originally marked eligible, it could
+      // be a copy of one. If that's the case, we inserted a primitive
+      // into its body.
+
+      // TODO: sometimes, this loop can be made no longer eligible by
+      // other transformations. Is that okay?
```

Comment:
Hypothetical, or do you have a real case where that may happen?

Reply:
I have a real case where this happened. I don't currently recall why, but I can try to find out.

Reply:
The case you bumped into might be LICM changing things to be non-eligible, because that's the only pass that happens between the first check and this one.

I am glad that you're protecting against it here, but at the same time, I am a bit worried that something like that can happen in general during LICM. So, checking it is not a prerequisite for this PR to be merged, but a curiosity in general.